### ISSUE TITLE
question about SGD one step

### ISSUE THREAD

**Issue by chen-chunling** [KarhouTam/Per-FedAvg#3]
In the section 5, the author wrote "Note that the model obtained by any of these three methods is later updated using one step of stochastic gradient descent at the test time". 
May I ask you why the **testset data** is used here instead of the **validation set data** ? This problem confuses me a lot.

**Comment by KarhouTam** [comment]
Hi, Chunling. First, thanks for your interest in my reproduction.
About your question, sorry, I just don't get it.😂 In `PerFedAvgClient.pers_N_eval()` I first evaluate the global model using the clients' validation set; then I sample a batch of data from the clients' training set and use it to fine-tune the model; finally I evaluate the fine-tuned model using the clients' validation set.
And I don't split each client's dataset into [train, val, test] (only train and val).
So I don't get your confusion. Can you describe your question more specifically?

**Comment by chen-chunling** [comment]
Thanks for your reply. Sorry about I didn't express the question clearly last time.

Firstly, let me recap the basic concepts in case there is a deviation in understanding later.
The dataset in each client is split into **[train, val]**. The **val set** is used to evaluate the performance of the fine-tuned model, it acts like the test set in other algorithms if we split the dataset into [train, val, test].  (If I am wrong here, please let me know)

Secondly, in the evaluation of perFedAvg, we need to update tilde{w} using val set mentioned above. The author of perFedAvg wrote 
> "_we take the output of the FedAvg method, and update it with one step of stochastic gradient descent with respect to the test data, and then evaluate its performance_". 

(Refer to Section 5. Numerical Experiments in their paper.) However, I did not see this procedure in your code  [`class PerFedAvgClient`](https://github.com/KarhouTam/Per-FedAvg/blob/14d964525549318920059717a60fbdb41133a938/perfedavg.py). 

Lastly, another implemention of perFedAvg I saw on github is [FedBABU](https://github.com/jhoon-oh/FedBABU/blob/master/main_per_fedavg.py), I am confused about the test data used in SGD one step update in the lines 96-99. Although the author rollback the model after evaluation (Lines 135-136), I'm curious about if we could use the test data in SGD one step here. If so, will we overfit in the test data?

**Comment by KarhouTam** [comment]
I think I get it. Your question is about the source of data that using for fine-tuning phase, right? 

I review the source code I got from the author(Split the dataset into [train, val] also.), and find that PerFedAvg actually uses data from the test(val) set to fine-tune the model. And in my reproduction, uses data from the trainset instead. That's the difference. And I think we won't overfit in the val(test) data, cause we just fine-tune the model using one data batch in one round.

But, in my opinion, using data from the val(test) set to fine-tune the model will let the model pre-fit the data, and performs better than it should. And it is like cheating in some way, I think? 😂 But it's ok to change my code by yourself. Just change the codes at `PerFedAvg.py`(Lines 185-187) to 
```
iter_valloader = iter(self.valloader)
x, y = self.get_data_batch(self.valloader, iter_valloader, self.device)
```
If my answer doesn't solve your problem actually, sorry for my misunderstanding.

P.S. About the source code, sorry I can't share it with you directly. Cause I don't know if the author wants it to be shared.

**Comment by chen-chunling** [comment]
Thank you very much for answering my question, you have solved my doubt, wish you the best of luck :) @KarhouTam
